### ISSUE TITLE
Bug Report: Gen4 planner generates inefficient (and potentially incorrect) query when joining multiple tables 

### ISSUE THREAD

**Issue by evaccaro** [vitessio/vitess#10819]
### Overview of the Issue

In our use case, we have three tables, one of them acting as a join table between the other two tables. The join is performed via unique lookup vindex columns. The V3 planner handles this type of query correctly (sends the full query to the target shard), but the Gen4 planner seems to get confused and breaks up the query into two parts. 

### `vschema.json`

```json
{
  "sharded": true,
  "vindexes": {
    "hash": {
      "type": "hash"
    },
    "issues_id_ks_idx": {
      "type": "lookup_test",
      "owner": "issues"
    },
    "issues_labels_id_ks_idx": {
      "type": "lookup_test",
      "owner": "issues_labels"
    },
    "labels_id_ks_idx": {
      "type": "lookup_test",
      "owner": "labels"
    }
  },
  "tables": {
    "issues": {
      "column_vindexes": [
        {
          "column": "repository_id",
          "name": "hash"
        },
        {
          "column": "id",
          "name": "issues_id_ks_idx"
        }
      ]
    },
    "issues_labels": {
      "column_vindexes": [
        {
          "column": "repository_id",
          "name": "hash"
        },
        {
          "column": "id",
          "name": "issues_labels_id_ks_idx"
        },
        {
          "column": "issue_id",
          "name": "issues_id_ks_idx"
        },
        {
          "column": "label_id",
          "name": "labels_id_ks_idx"
        }
      ]
    },
    "labels": {
      "column_vindexes": [
        {
          "column": "repository_id",
          "name": "hash"
        },
        {
          "column": "id",
          "name": "labels_id_ks_idx"
        }
      ]
    }
  }
}
```

### Query

```sql
SELECT COUNT(*) 
FROM `issues` 
INNER JOIN `issues_labels` ON `issues_labels`.`issue_id` = `issues`.`id` 
INNER JOIN `labels` ON `labels`.`id` = `issues_labels`.`label_id` 
WHERE `issues`.`repository_id` = 143 AND (issues_labels.label_id = '152')
```

### V3 Query Plan

```json
{
    "QueryType": "SELECT",
    "Original": "SELECT COUNT(*) FROM `issues` INNER JOIN `issues_labels` ON `issues_labels`.`issue_id` = `issues`.`id` INNER JOIN `labels` ON `labels`.`id` = `issues_labels`.`label_id` WHERE `issues`.`repository_id` = 143 AND (issues_labels.label_id = '152')",
    "Instructions": {
    "OperatorType": "Route",
    "Variant": "EqualUnique",
    "Keyspace": {
        "Name": "issues",
        "Sharded": true
    },
    "FieldQuery": "select count(*) from issues join issues_labels on issues_labels.issue_id = issues.id join labels on labels.id = issues_labels.label_id where 1 != 1",
    "Query": "select count(*) from issues join issues_labels on issues_labels.issue_id = issues.id join labels on labels.id = issues_labels.label_id where issues.repository_id = 143 and issues_labels.label_id = '152'",
    "Table": "issues, issues_labels, labels",
    "Values": [
        "INT64(143)"
    ],
    "Vindex": "hash"
    }
}
```

### Gen4 Query Plan

```json
{
  "QueryType": "SELECT",
  "Original": "SELECT COUNT(*) FROM `issues` INNER JOIN `issues_labels` ON `issues_labels`.`issue_id` = `issues`.`id` INNER JOIN `labels` ON `labels`.`id` = `issues_labels`.`label_id` WHERE `issues`.`repository_id` = 143 AND (issues_labels.label_id = '152')",
  "Instructions": {
    "OperatorType": "Aggregate",
    "Variant": "Scalar",
    "Aggregates": "sum_count_star(0) AS count(*)",
    "Inputs": [
      {
        "OperatorType": "Projection",
        "Expressions": [
          "[COLUMN 0] * [COLUMN 1] as count(*)"
        ],
        "Inputs": [
          {
            "OperatorType": "Join",
            "Variant": "Join",
            "JoinColumnIndexes": "L:1,R:1",
            "JoinVars": {
              "issues_id": 0
            },
            "TableName": "issues_issues_labels, labels",
            "Inputs": [
              {
                "OperatorType": "Route",
                "Variant": "EqualUnique",
                "Keyspace": {
                  "Name": "issues",
                  "Sharded": true
                },
                "FieldQuery": "select issues.id, count(*), weight_string(issues.id) from issues where 1 != 1 group by issues.id, weight_string(issues.id)",
                "Query": "select issues.id, count(*), weight_string(issues.id) from issues where issues.repository_id = 143 group by issues.id, weight_string(issues.id)",
                "Table": "issues",
                "Values": [
                  "INT64(143)"
                ],
                "Vindex": "hash"
              },
              {
                "OperatorType": "Route",
                "Variant": "EqualUnique",
                "Keyspace": {
                  "Name": "issues",
                  "Sharded": true
                },
                "FieldQuery": "select 1, count(*) from issues_labels, labels where 1 != 1 group by 1",
                "Query": "select 1, count(*) from issues_labels, labels where issues_labels.label_id = '152' and issues_labels.issue_id = :issues_id and labels.id = issues_labels.label_id group by 1",
                "Table": "issues_labels, labels",
                "Values": [
                  "VARCHAR(\"152\")"
                ],
                "Vindex": "labels_id_ks_idx"
              }
            ]
          }
        ]
      }
    ]
  },
  "TablesUsed": [
    "issues.issues",
    "issues.issues_labels",
    "issues.labels"
  ]
}
```

### Binary Version

```sh
Version: 15.0.0-SNAPSHOT (Git revision 665cb4139ce2cc705b7d8d99b7c760f37a285ee1 branch 'main') built on Mon Jun 27 10:10:31 UTC 2022 by codespace@codespaces-7bc63a using go1.18.3 linux/amd64
```

### Operating System and Environment details

```sh
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.4 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.4 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

**Comment by arthurschreiber** [comment]
We'd expect the constraint on `repository_id` (a `hash` vindex) to be pushed through the `issues_labels` join down to the join with `labels` as well. Instead it looks like the planner thinks that it has to use the unique lookup vindex `labels_id_ks_idx` on `issues_labels.label_id` instead, and thus breaks up the query into two parts.

Removing the condition on `issues.repository_id` generates the following plan:

```json
{
  "QueryType": "SELECT",
  "Original": "SELECT COUNT(*) FROM `issues` INNER JOIN `issues_labels` ON `issues_labels`.`issue_id` = `issues`.`id` INNER JOIN `labels` ON `labels`.`id` = `issues_labels`.`label_id` WHERE (issues_labels.label_id = '152')",
  "Instructions": {
    "OperatorType": "Route",
    "Variant": "EqualUnique",
    "Keyspace": {
      "Name": "issues",
      "Sharded": true
    },
    "FieldQuery": "select count(*) from issues, issues_labels, labels where 1 != 1",
    "Query": "select count(*) from issues, issues_labels, labels where issues_labels.label_id = '152' and issues_labels.issue_id = issues.id and labels.id = issues_labels.label_id",
    "Table": "issues, issues_labels, labels",
    "Values": [
      "VARCHAR(\"152\")"
    ],
    "Vindex": "labels_id_ks_idx"
  }
}
```

In this case, the planner understands that it can push the constraint from `issues_labels.label_id` to both joins.

When multiple constraints are given, the planner should pick the "cheapest" one (in this case, the `hash` vindex) and push it down to all tables that are joined via unique lookup vindex columns.

**Comment by glortho** [comment]
Fixed by https://github.com/vitessio/vitess/pull/10942

**Comment by frouioui** [comment]
This issue seems to be fixed by #10942 indeed. When running the same query with the given VSchema the Gen4 plan is:

```json
{
  "QueryType": "SELECT",
  "Original": "SELECT COUNT(*) FROM `issues` INNER JOIN `issues_labels` ON `issues_labels`.`issue_id` = `issues`.`id` INNER JOIN `labels` ON `labels`.`id` = `issues_labels`.`label_id` WHERE `issues`.`repository_id` = 143 AND (issues_labels.label_id = '152')",
  "Instructions": {
	"OperatorType": "Route",
	"Variant": "EqualUnique",
	"Keyspace": {
	  "Name": "test",
	  "Sharded": true
	},
	"FieldQuery": "select count(*) from issues, issues_labels, labels where 1 != 1",
	"Query": "select count(*) from issues, issues_labels, labels where issues.repository_id = 143 and issues_labels.label_id = '152' and issues_labels.issue_id = issues.id and labels.id = issues_labels.label_id",
	"Table": "issues, issues_labels, labels",
	"Values": [
	  "INT64(143)"
	],
	"Vindex": "hash"
  },
  "TablesUsed": [
	"test.issues",
	"test.issues_labels",
	"test.labels"
  ]
}
```

We are using the `hash` vindex which is shared by all three tables.
